File: near/src/main/java/com/adroitandroid/near/model/Host.java
package com.adroitandroid.near.model;

/**
 * Created by pv on 20/06/17.
 */

import android.os.Parcel;
import android.os.Parcelable;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import org.jetbrains.annotations.Contract;

import java.net.InetAddress;
import java.net.UnknownHostException;

public class Host implements Parcelable {

    private final InetAddress inetAddress;
    private final String name;

    private Host() {
        inetAddress = InetAddress.getLoopbackAddress();
        name = "dummy";
    }

    public Host(InetAddress address, String name) {
        this.inetAddress = address;
        this.name = name;
    }

    protected Host(@NonNull Parcel in) {
        this.name = in.readString();
        this.inetAddress = getInetAddressFrom(in);
    }

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(@NonNull Parcel dest, int flags) {
        dest.writeString(name);
        dest.writeInt(inetAddress.getAddress().length);
        dest.writeByteArray(inetAddress.getAddress());
    }

    public static final Creator<Host> CREATOR = new Creator<Host>() {

        @NonNull
        @Contract("_ -> new")
        @Override
        public Host createFromParcel(Parcel in) {
            return new Host(in);
        }


        @NonNull
        @Contract(value = "_ -> new", pure = true)
        @Override
        public Host[] newArray(int size) {
            return new Host[size];
        }
    };


    @Nullable
    private InetAddress getInetAddressFrom(@NonNull Parcel in) {
        byte[] addr = new byte[in.readInt()];
        in.readByteArray(addr);
        try {
            return InetAddress.getByAddress(addr);
        } catch (UnknownHostException e) {
            e.printStackTrace();
            return null;
        }
    }

    public String getHostAddress() {
        return inetAddress.getHostAddress();
    }

    public String getName() {
        return name;
    }

    @Contract(value = "null -> false", pure = true)
    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (!(o instanceof Host)) return false;

        Host host = (Host) o;

        return inetAddress.getHostAddress().equals(host.inetAddress.getHostAddress());
    }

    @Override
    public int hashCode() {
        return inetAddress.hashCode();
    }

    @Override
    public String toString() {
        return "Host{" +
                "address=" + inetAddress.getHostAddress() +
                ", name='" + name + '\'' +
                '}';
    }
}

File: near/src/main/java/com/adroitandroid/near/model/Host.kt
package com.adroitandroid.near.model

import android.os.Parcel
import android.os.Parcelable
import org.json.JSONObject
import java.net.InetAddress
import java.net.UnknownHostException

open class Host : Parcelable {
    private val inetAddress: InetAddress
    val name: String
    val filterText: String

    private constructor() {
        inetAddress = LOOPBACK_ADDRESS
        name = DUMMY
        filterText = FILTER_TEXT
    }

    constructor(address: InetAddress, name: String, filterText: String) {
        inetAddress = address
        this.name = name
        this.filterText = filterText
    }

    protected constructor(parcel: Parcel) {
        name = parcel.readString() ?: DUMMY
        inetAddress = getInetAddressFrom(parcel) ?: LOOPBACK_ADDRESS
        filterText = parcel.readString() ?: FILTER_TEXT
    }

    override fun describeContents(): Int {
        return 0
    }

    override fun writeToParcel(dest: Parcel, flags: Int) {
        dest.writeString(name)
        dest.writeInt(inetAddress.address.size)
        dest.writeByteArray(inetAddress.address)
        dest.writeString(filterText)
    }

    private fun getInetAddressFrom(parcel: Parcel): InetAddress? {
        val addr = ByteArray(parcel.readInt())
        parcel.readByteArray(addr)
        return try {
            InetAddress.getByAddress(addr)
        } catch (e: UnknownHostException) {
            e.printStackTrace()
            null
        }
    }

    val hostAddress: String
        get() = inetAddress.hostAddress

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is Host) return false
        return inetAddress.hostAddress == other.inetAddress.hostAddress
    }

    override fun hashCode(): Int {
        return inetAddress.hashCode()
    }

    override fun toString(): String {
        val hostMap: Map<String, Any> = mapOf("address" to inetAddress.hostAddress,
                JSON_NAME to name,
                JSON_FILTER_TEXT to filterText)
        return JSONObject(hostMap).toString()
    }

    companion object CREATOR : Parcelable.Creator<Host> {
        const val DUMMY = "dummy"
        val LOOPBACK_ADDRESS: InetAddress = InetAddress.getLoopbackAddress()
        const val FILTER_TEXT = ""
        const val JSON_NAME = "name"
        const val JSON_FILTER_TEXT = "filterText"

        override fun createFromParcel(parcel: Parcel): Host {
            return Host(parcel)
        }

        override fun newArray(size: Int): Array<Host?> {
            return arrayOfNulls(size)
        }
    }
}